packages/flutter/lib/src/rendering/proxy_box.dart

```diff
@@ -994,6 +994,37 @@ class RenderClipOval extends _RenderCustomClip<Rect> {
   }
 }
 
+/// Clips its child using a path.
+///
+/// Takes a callback which, during paint, returns a path that should be used to
+/// prevent the child from painting outside the path.
+class RenderClipPath extends _RenderCustomClip<Path> {
+  RenderClipPath({
+    RenderBox child,
+    CustomClipper<Path> clipper
+  }) : super(child: child, clipper: clipper);
+
+  Path _cachedPath;
+
+  @override
+  Path get _defaultClip => new Path()..addRect(Point.origin & size);
+
+  @override
+  bool hitTest(HitTestResult result, { Point position }) {
+    if (_cachedPath == null || !_cachedPath.contains(position))
+      return false;
+    return super.hitTest(result, position: position);
+  }
+
+  @override
+  void paint(PaintingContext context, Offset offset) {
+    if (child != null) {
+      _cachedPath = _clip; // invokes the callback
```

Comment:
I've updated all the clip render objects to cache the results of calling the delegate at layout time.